### PR TITLE
improving TimmBackbone to support FrozenBatchNorm2d

### PR DESCRIPTION
**Pull request by rafaelpadilla**
# What does this PR do?

The `TimmBackbone` offers various customizable parameters, such as `num_channels`, `features_only`, and `out_indices`.

Currently, we lack direct support for timm's `freeze_batch_norm_2d(...) `function, which substitutes the backbone's `BatchNorm2d` layers with `FrozenBatchNorm2d`.

For models like RTDetr, Detr, ConditionalDetr, DeformableDetr, etc that require freezing of batch norm layers, one currently needs to either reimplement and modify the backbone manually or use the `timm.layers.freeze_batch_norm_2d(...)` function, like so:

```python
my_backbone = TimmBackbone.from_pretrained("resnet50d")
my_backbone = timm.layers.freeze_batch_norm_2d(my_backbone)
```
To simplify this, this PR introduces the `TimmBackboneConfig.freeze_batch_norm_2d` parameter, which defaults to False for backward compatibility. When creating the backbone, if this parameter is `True`, it will apply `timm's freeze_batch_norm_2d(...) ` function.

This update simplifies the process of creating backbone models. For models requiring a switch from `BatchNorm2d` to `FrozenBatchNorm2d`, the `TimmBackbone` can be used directly, eliminating additional steps.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@amyeroberts

### CODE REVIEW

File: src/transformers/models/timm_backbone/modeling_timm_backbone.py [hunk 1]
```diff
@@ -72,6 +72,11 @@ def __init__(self, config, **kwargs):
             out_indices=out_indices,
             **kwargs,
         )
+
+        # Converts all `BatchNorm2d` and `SyncBatchNorm` or `BatchNormAct2d` and `SyncBatchNormAct2d` layers of provided module into `FrozenBatchNorm2d` or `FrozenBatchNormAct2d` respectively
+        if hasattr(config, "freeze_batch_norm_2d") and config.freeze_batch_norm_2d is True:
+            self._backbone = timm.layers.freeze_batch_norm_2d(self._backbone)
```
Comment (amyeroberts) [inline src/transformers/models/timm_backbone/modeling_timm_backbone.py:78]: Rather than have this hard coded within the `__init__` it would be good to have both the freeze and unfreeze methods than can be applied to the model as-needed i.e.

```
class TimmBackbone(PreTrainedModel, BackboneMixin):
    def __init__(...):
        if getattr(config, "freeze_batch_norm_2d", False):
            self.freeze_batch_norm_2d()

    def freeze_batch_norm_2d(self):
        self._backbone = timm.layers.freeze_batch_norm_2d(self._backbone)

    def unfreeze_bach_norm_2d(self):
        timm.layers.unfreeze_batch_norm_2d(self._backbone)
```

**Review by amyeroberts — COMMENTED** [review]
Nice! 

Great addition to the backbone class. Just a comment on adding it as a method instead and including the unfreeze option too.

File: src/transformers/models/timm_backbone/modeling_timm_backbone.py (hunk 1 above)
Comment (rafaelpadilla) [reply src/transformers/models/timm_backbone/modeling_timm_backbone.py]: Done! 
I like the idea of having our own freeze and unfreeze methods. :)
I tested and noticed that `freeze_batch_norm_2d(...)` and `unfreeze_batch_norm_2d(...)` are both done in place, so no need to return.

File: src/transformers/models/timm_backbone/modeling_timm_backbone.py [hunk 2]
```diff
@@ -72,6 +72,13 @@ def __init__(self, config, **kwargs):
             out_indices=out_indices,
             **kwargs,
         )
+
+        # Converts all `BatchNorm2d` and `SyncBatchNorm` or `BatchNormAct2d` and `SyncBatchNormAct2d` layers of provided module into `FrozenBatchNorm2d` or `FrozenBatchNormAct2d` respectively
+        if getattr(config, "freeze_batch_norm_2d", False):
+            self.freeze_batch_norm_2d()
+        else:
+            self.unfreeze_batch_norm_2d()
```
Comment (amyeroberts) [inline src/transformers/models/timm_backbone/modeling_timm_backbone.py:80]: I'm not sure that we want to do `unfreeze_batch_norm_2d` automatically if the argument isn't in the config. If I loaded in a saved backbone, which previously did not have this argument in its config but did have frozen batch norm layers - this will now unfreeze them silently. Moreover, it's not obvious that `freeze_batch_norm_2d=False` unfreezes, it can equally be read as just not freezing. 

Instead, I would only apply if the argument is positive, and let the user unfreeze directly with `model.unfreeze_batch_norm_2d` if they need.  

```suggestion        
        # Converts all `BatchNorm2d` and `SyncBatchNorm` or `BatchNormAct2d` and `SyncBatchNormAct2d` layers of provided module into `FrozenBatchNorm2d` or `FrozenBatchNormAct2d` respectively
        if getattr(config, "freeze_batch_norm_2d", False):
            self.freeze_batch_norm_2d()
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Just a note on the backwards compatibility of the model when loading in pretrained checkpoints.

File: src/transformers/models/timm_backbone/modeling_timm_backbone.py (hunk 2 above)
Comment (rafaelpadilla) [reply src/transformers/models/timm_backbone/modeling_timm_backbone.py:80]: makes total sense :)

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this!
